rust/src/cliwrap/yumdnf.rs

```diff
@@ -0,0 +1,152 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use anyhow::{anyhow, Result};
+use indoc::indoc;
+use std::os::unix::process::CommandExt;
+use std::process::Command;
+use structopt::StructOpt;
+
+const UNHANDLED: &str = indoc::indoc! {r#"
+rpm-ostree: Unknown arguments passed to yum/dnf compatibility wrapper.
+
+Please see `yum --help` for currently accepted commands via this
+wrapper, and `rpm-ostree --help` for more information about.the rpm-ostree
+project.
+"#};
+
+#[derive(Debug, StructOpt)]
+#[structopt(
+    name = "yumdnf-rpmostree",
+    about = "Compatibility wrapper implementing subset of yum/dnf CLI"
+)]
+#[structopt(rename_all = "kebab-case")]
+/// Main options struct
+enum Opt {
+    /// Start an upgrade of the operating system
+    Upgrade,
+    /// Start an upgrade of the operating system
+    Update,
+    /// Display information about system state
+    Status,
+    /// Perform a search of packages.
+    Search {
+        /// Search terms
+        terms: Vec<String>,
+    },
+    /// Will return an error suggesting other approaches.
+    Install {
+        /// Set of packages to install
+        packages: Vec<String>,
+    },
+    /// Layer RPM packages as operating system extensions
+    InstallExtension {
+        /// Set of packages to install
+        packages: Vec<String>,
+    },
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum RunDisposition {
+    HelpOrVersionDisplayed,
+    ExecRpmOstree(Vec<String>),
+    UseSomethingElse(&'static str),
+    Unhandled,
+}
+
+fn disposition(argv: &[&str]) -> Result<RunDisposition> {
+    let opt = match Opt::from_iter_safe(std::iter::once(&"yum").chain(argv.iter())) {
+        Ok(v) => v,
+        Err(e)
+            if e.kind == clap::ErrorKind::VersionDisplayed
+                || e.kind == clap::ErrorKind::HelpDisplayed =>
+        {
+            return Ok(RunDisposition::HelpOrVersionDisplayed)
+        }
+        Err(_) => {
+            return Ok(RunDisposition::Unhandled);
+        }
+    };
+
+    let disp = match opt {
+        Opt::Upgrade | Opt::Update => RunDisposition::ExecRpmOstree(vec!["upgrade".into()]),
+        Opt::Status => RunDisposition::ExecRpmOstree(vec!["status".into()]),
+        Opt::Install { packages: _ } => {
+            // TODO analyze packages to find e.g. `gcc` (not ok, use `toolbox`) versus `libvirt` (ok)
+            RunDisposition::UseSomethingElse(indoc! {r##"
+This system is rpm-ostree based.  Before installing packages to
+the host root filesystem, consider other options:
```

Comment:
And this part clearly needs to be customizable by the OS; doesn't make sense to mention `flatpak` on CoreOS, and some operating systems may want to swap out things.

Am currently thinking of a file like `/usr/share/rpm-ostree/install-message.txt` or something that we just open and display.  But we could also pipe the package names to a child process as a plugin so it could say "oh you're asking for gcc, use `toolbox`"?

Reply:
Yeah, a way to customize the message makes a lot of sense. That way they can each also link to their respective documentation. (Though we could probably have some generic documentation upstream we default to.)